src/main/ruby/core/process_mirror.rb

```diff
@@ -175,84 +183,84 @@ def initialize(env_or_cmd, *args)
+
+                  # When redirecting multiple fds to one file, as in
+                  #
+                  # [:out, :err] => "/some/file"
+                  #
+                  # We want this to be equivalent to
+                  #
+                  # :hidden_fd => "/some/file", :out => :hidden_fd, :err => :hidden_fd
+                  #
+                  # opening the destination once, and only then
+                  # redirecting everything to that destination.
+                  #
+                  # Therefore we open all files in the parent process,
+                  # and only specify redirection for the child process.
+                  #
+                  # We don't use open actions for this as we want to
+                  # guarantee the order and need to work around a bug in
+                  # old versions of glibc which do not correctly
+                  # duplicate the string containing the file path.
+
+                  fds = key.map { |k| convert_io_fd(k) }
+                  to = convert_to_fd value, fds.first
+                  fds.each { |fd| redirect @options, fd, to }
+                when :unsetenv_others
+                  if value
+                    @options[:unsetenv_others] = true
+                  end
+                when :pgroup
+                  if value == true
+                    value = 0
+                  elsif value
+                    value = Rubinius::Type.coerce_to value, ::Integer, :to_int
+                    raise ArgumentError, "negative process group ID : #{value}" if value < 0
+                  else
+                    value = -1
+                  end
+                  @options[key] = value
+                when :chdir
+                  @options[key] = Rubinius::Type.coerce_to_path(value)
+                when :umask
+                  @options[key] = value
+                when :close_others
+                  @options[key] = value if value
                 else
-                  value = -1
+                  raise ArgumentError, "unknown exec option: #{key.inspect}"
                 end
-                @options[key] = value
-              when :chdir
-                @options[key] = Rubinius::Type.coerce_to_path(value)
-              when :umask
-                @options[key] = value
-              when :close_others
-                @options[key] = value if value
-              else
-                raise ArgumentError, "unknown exec option: #{key.inspect}"
               end
             end
-          end
 
-          if env
-            array = (@options[:env] ||= [])
+            if env
+              array = (@options[:env] ||= [])
 
-            env.each do |key, value|
-              array << [convert_env_key(key), convert_env_value(value)]
+              env.each do |key, value|
+                array << [convert_env_key(key), convert_env_value(value)]
+              end
             end
+          rescue
+            close_files
+            raise
```

Comment:
Could you extract most of the logic of initialize and put it in a method `parse`?
Then we can just use the ensure in the caller.

Reply:
@aardvark179 I did the split and a bit of cleanup.